text/0000-constrained-naked.md

```diff
@@ -0,0 +1,142 @@
+
+# Motivation
+
+Naked functions have long been a feature of compilers. These functions are typically defined as normal functions in every regard, except that the compiler does not emit the function prologue and epilogue. Rust's early attempt to support this feature ([RFC 1201](https://github.com/rust-lang/rfcs/blob/master/text/1201-naked-fns.md)) mostly copied the existing compiler behaviors.
+
+However, naked functions are often avoided in practice because their behavior is not well defined. The root cause of this problem is that naked functions are defined by negation: they are functions which lack a prologue and epilogue. Unfortunately, functions that lack a prologue and epilogue present a number of complicated problems that the compiler needs to solve and developers need to work around. And there is a long history of compilers and developers getting this wrong.
+
+This document seeks to define naked functions in a much more constrained, positivistic way. In doing so, naked functions can become more useful.
+
+# Naked function definition
+
+A naked function is a type of FFI function with a defined calling convention and a body which contains only assembly code which can rely upon the defined calling convention.
+
+A naked function is identified by the `#[naked]` attribute and:
+1. must specify a calling convention besides `extern "Rust"`.
+1. must define only FFI-safe arguments and return types.
+1. must not specify the `#[inline]` or `#[inline(always)]` attribute.
+1. must have a body which contains only a single `asm!()` statement which:
+    1. may be wrapped in an `unsafe` block.
+    1. must not contain any operands except `const` or `sym`.
+    1. must contain the `noreturn` option.
+    1. must not contain any other options except `att_syntax`.
+    1. must ensure that the requirements of the calling convention are followed.
+
+In exchange for the above constraints, the compiler commits to:
+1. produce a clear error if any of the above rules are violated.
+1. never inline the function (implicit `#[inline(never)]`).
+1. emit no instructions to the function body that are not contained in the `asm!()` statement.
+
+As a (weaker) correlary to the last compiler commitment, since the compiler generates no additional instructions, the initial state of all registers in the `asm!()` statement conform to the specified calling convention.
+
+# Custom Calling Convention
+
+In order to expand the usefulness of naked functions, this document also define a new calling convention, `extern "custom"`, that can be used only with naked functions. Functions with `extern "custom"` are not callable from Rust and must be marked as`unsafe`. The function definition must contain no arguments or return type. For example:
```

Comment:
I don't see why we should require such functions to be `unsafe`. They can't be called by Rust code at all so it shouldn't matter whether they are safe or unsafe.

Reply:
I was originally going to say the same thing, but it actually somehow makes sense: these functions always have a caller requirement (using the right calling convention), so it's kind of logical that passing around `extern "custom" fn()` pointers should be forbidden to force the use of `unsafe extern "custom" fn()`.

This being said, it is totally possible to imagine a custom calling convention whereby eg. the argument is passed in a global variable and all bitpatterns are valid, so any call would actually be valid, and thus it wouldn't require `unsafe`.

So, with all the above considered, I also believe that it'd be best to remove this requirement, that AFAIU doesn't bring much additional value.

Reply:
I added that merely to be defensive, though I had the same thought. I'm curious what others think about this.